Makefile

```diff
@@ -128,69 +129,141 @@ acncli: acncli-binary acncli-archive
 	cd $(NPM_DIR) && CGO_ENABLED=0 go build -v -o $(NPM_BUILD_DIR)/azure-npm$(EXE_EXT) -ldflags "-X main.version=$(VERSION) -X $(NPM_AI_PATH)=$(NPM_AI_ID)" -gcflags="-dwarflocationlists=true"
 
 
-########################### Container Images ########################
+##@ Containers
+
+# prefer buildah, if available, but fall back to docker if that binary is not in the path.
+CONTAINER_BUILDER=buildah
+ifeq (, $(shell which $(CONTAINER_BUILDER)))
+CONTAINER_BUILDER=docker
+endif
+
+containerize-buildah: # util target to build container images using buildah. do not invoke directly.
+	buildah build \
+		--jobs 16 \
+		--platform $(PLATFORM) \
+		-f $(DOCKERFILE) \
+		--build-arg VERSION=$(VERSION) $(EXTRA_BUILD_ARGS) \
+		-t $(REGISTRY)/$(IMAGE):$(TAG) \
+		.
+
+containerize-docker: # util target to build container images using docker buildx. do not invoke directly.
+	docker buildx build \
+		--platform $(PLATFORM) \
+		-f $(DOCKERFILE) \
+		--build-arg VERSION=$(VERSION) $(EXTRA_BUILD_ARGS) \
+		-t $(REGISTRY)/$(IMAGE):$(TAG) \
+		.
+
+container-tag-test: # util target to retag an image with -test suffix. do not invoke directly.
+	$(CONTAINER_BUILDER) tag \
+		$(REGISTRY)/$(IMAGE):$(TAG) \
+		$(REGISTRY)/$(IMAGE):$(TAG)-test
+
+container-push: # util target to publish container image. do not invoke directly.
+	$(CONTAINER_BUILDER) push \
+		$(REGISTRY)/$(IMAGE):$(TAG)
+
+container-pull: # util target to pull container image. do not invoke directly.
+	$(CONTAINER_BUILDER) pull \
+		$(REGISTRY)/$(IMAGE):$(TAG)
+
+container-info: # util target to write container info file. do not invoke directly.
+	# these commands need to be root due to some ongoing perms issues in the pipeline.
+	sudo mkdir -p $(IMAGE_DIR) 
+	sudo chown -R $$(whoami) $(IMAGE_DIR) 
+	sudo chmod -R 777 $(IMAGE_DIR)
+	echo $(IMAGE):$(TAG) > $(IMAGE_DIR)/$(FILE)
+
+azure-cns-image: ## build azure-cns container image.
+	$(MAKE) containerize-$(CONTAINER_BUILDER) \
+		PLATFORM=$(GOOS)/$(GOARCH) \
+		DOCKERFILE=cns/Dockerfile \
+		REGISTRY=$(IMAGE_REGISTRY) \
+		IMAGE=$(AZURE_CNS_IMAGE) \
+		EXTRA_BUILD_ARGS='--build-arg CNS_AI_PATH=$(CNS_AI_PATH) --build-arg CNS_AI_ID=$(CNS_AI_ID)' \
+		TAG=$(VERSION)
+	$(MAKE) container-info IMAGE=$(AZURE_CNS_IMAGE) TAG=$(VERSION) FILE=$(CNS_IMAGE_INFO_FILE)
 
 # Build the tools images
-.PHONY: tools-images
-tools-images:
-	$(MKDIR) $(IMAGE_DIR)
-	docker build --no-cache -f ./tools/acncli/Dockerfile --build-arg VERSION=$(VERSION) -t $(AZURE_CNI_IMAGE):$(VERSION) .
+tools-images: ## build the acncli container image.
+	$(MAKE) containerize-docker \
+		PLATFORM=$(GOOS)/$(GOARCH) \
+		DOCKERFILE=tools/acncli/Dockerfile \
+		REGISTRY=$(IMAGE_REGISTRY) \
+		IMAGE=$(AZURE_CNI_IMAGE) \
+		TAG=$(VERSION)
 	docker save $(AZURE_CNI_IMAGE):$(VERSION) | gzip -c > $(IMAGE_DIR)/$(CNI_IMAGE_ARCHIVE_NAME)
 
+
+# Build the Azure NPM image.
+azure-npm-image: ## build the azure-npm container image.
+	$(MAKE) containerize-$(CONTAINER_BUILDER) \
+			PLATFORM=$(GOOS)/$(GOARCH) \
```

Comment:
I like the idea, is there a way to specify multi-arch when `$(CONTAINER_BUILDER)` is `docker`?
Trying to preserve this https://github.com/Azure/azure-container-networking/blob/7fbbd1e0b7aaa35235b5e8fc2f66d39d0d9ce329/Makefile#L99

Reply:
yes, we could invoke the containerize-* targets with `PLATFORM=linux/amd64,linux/arm64` flag, and this would get passed to the builder's `--platform` flag. the behavior would be the same as the current buildx targets.
however, the intent of this PR is to set us up to go in a different direction from the current serial multiarch buildx builds - we want to set up separate build stages for every platform target so that they can run very parallelized (ex: npm arm64 and amd64 can build concurrently and both get their own full, independent, arch-specific tests run).

Reply:
i have updated the pipeline and makefile to _do_ what i said previously - it now runs the individual image x os x arch builds in parallel then composes and pushes the multiplatform manifest in a subsequent next step